### ISSUE TITLE
Differences between get_test_eval and prediction on test set

### ISSUE THREAD

**Issue by ctlaltdefeat** [catboost/catboost#125]
For some reason, after fitting a model with a test set,  the array returned by get_test_eval() and the prediction of the model on the same test set are quite different (the actual prediction has a worse evaluation metric).

Any idea why that could be happening? I am using use_best_model=True.

**Comment by annaveronika** [comment]
Could you provide the code that you are running?

**Comment by ctlaltdefeat** [comment]
It seems to occur only when using categorical features. Here's sample code:

```
import pandas as pd
from catboost import CatBoostRegressor
from sklearn.model_selection import train_test_split
import numpy as np

c_params = {
    'calc_feature_importance': True,
    'depth': 4,
    'iterations': 100,
    'learning_rate': 0.1,
}

train = pd.DataFrame(
    {'x1': np.round(np.random.rand(1, 10000)[0] * 500), 'x2': np.random.rand(1, 10000)[0],
     'y': np.random.rand(1, 10000)[0]})
categorical_features = ['x1']
ind_t, ind_v = train_test_split(train.index, test_size=0.2, random_state=6)
cbm = CatBoostRegressor(**c_params, random_seed=13)
cbm.fit(
    X=train.iloc[ind_t].drop(['y'], axis=1),
    y=train.iloc[ind_t]['y'],
    eval_set=(train.iloc[ind_v].drop(['y'], axis=1),
              train.iloc[ind_v]['y']),
    cat_features=[
        i for i in range(len(train.iloc[ind_t].drop(['y'], axis=1).columns))
        if train.iloc[ind_t].drop(['y'], axis=1).columns[i] in categorical_features
    ],
)
print(cbm.get_test_eval()[0] - cbm.predict(train.iloc[ind_v].drop(['y'], axis=1)))
```
Returns 
```
[ 0.0004363  -0.00110495  0.0039695  ..., -0.01455476  0.00075446
  0.00464302]
```
for me using latest versions of python and modules.

Running the same thing except with ```categorical_features = []``` gives an array of 0s as expected.

**Comment by annaveronika** [comment]
Thank you for the report! This should be fixed in the last version, please try udpating catboost.

**Comment by ctlaltdefeat** [comment]
I am/was using 0.2.4.

**Comment by annaveronika** [comment]
please try updatinig to the next version, now the bug should be fixed.

**Comment by ctlaltdefeat** [comment]
Thanks, I'll try. 
Can I have some information about this? What caused this? What was fixed? Is this change likely to improve accuracy?
I just have some models that took a long time to train and I'm wondering if I should retrain with the new version.

**Comment by annaveronika** [comment]
We have updated the way we are storing the documents in pool, so that we use less memory. And we made a bug there that has caused wrong approx value calculation, so no, no retraining should be done.

About your models - it might happen that the new version of CatBoost will not open the old models, because at some point we made breaking changes to our model. If this happens you should try open them with the old version.

We are currently working on moving to flat buffers for our models, after that most probably no breaking changes to the models will be done.

**Comment by ctlaltdefeat** [comment]
Ah, ok.
That means that the prediction results of the exact predictions shouldn't change in the new version compared to the last, right?

**Comment by annaveronika** [comment]
The prediction results might change - we make some speedups that include changing summation order for example.

**Comment by kizill** [comment]
Some time ago we changed default way of "counter" CTR calculation: we update numerator and denominator both on learn and test by default, but we've forgotten to pass this new parameter to ModelCTR table calculation code.
So, your trained model will give you the same predictions with new code and new version will produce proper counter CTR values table in saved model.

**Comment by ctlaltdefeat** [comment]
Thank you!
